### PR TITLE
Add LeftShift and RightShift

### PR DESCRIPTION
**Pull request by sky58**
This commit doesn't pass the compile yet.
Since I'm not sure whether I should accept the boolean variable for the first argument of the shift operation or not, and how should I decide the type of return value from the type of arguments.
Could you decide how should I do?

### CODE REVIEW

File: chainerx_cc/chainerx/routines/binary.cc [hunk 1]
```diff
@@ -122,4 +142,28 @@ Array BitwiseXor(const Array& x1, Scalar x2) {
...
+    CheckBitwiseDtypes(x1, x2);
+    return internal::BroadcastBinary(BitwiseImpl<LeftShiftKernel>, x1, x2, ResultType(x1, x2));
+}
+
+Array LeftShift(const Array& x1, Scalar x2) {
+    CheckBitwiseDtypes(x1, x2);
+    return internal::Binary(BitwiseASImpl<LeftShiftASKernel>, x1, x2, ResultType(x1, x2));
+}
+
+Array LeftShift(Scalar x1, const Array& x2) { return LeftShift(x2, x1); }
```
Comment (kshitij12345) [inline chainerx_cc/chainerx/routines/binary.cc]: I guess this is wrong as `<<` is not commutative i.e. `2 << 5 != 5 << 2`.

It is ok for `&` or `|` because, `2&5 = 5&2`.

**Comment by kshitij12345** [comment]
> how should I decide the type of return value from the type of arguments.
Could you decide how should I do?

From what I have seen using `numpy`, if `array` is of `dtype=bool`, any bit-shift operation returns array of `int64`, while for other `dtypes` the returned array preserves the dtype of input array.

File: chainerx_cc/chainerx/routines/binary.cc (hunk 1 above)
Comment (sky58) [reply chainerx_cc/chainerx/routines/binary.cc]: Now I fix this bug. Is this solution ok for you?

**Comment by sky58** [comment]
> > how should I decide the type of return value from the type of arguments.
> > Could you decide how should I do?
> 
> From what I have seen using `numpy`, if `array` is of `dtype=bool`, any bit-shift operation returns array of `int64`, while for other `dtypes` the returned array preserves the dtype of input array.
> 
> In `chainerx`, you probably use,
> 
> https://github.com/chainer/chainer/blob/65bcf2fe8a0afe9c97ddf9db45d9a180f8ff29f0/chainerx_cc/chainerx/routines/math.cc#L398
> 
> to get the default `dtype` for the given kind.

I'm not sure your intention to show this example. If I follow this, it seems that dtype becoms float and it's not suitable for this situation, right?

**Comment by kshitij12345** [comment]
> I'm not sure your intention to show this example. If I follow this, it seems that dtype becoms float and it's not suitable for this situation, right?

Ah I guess I wasn't clear enough. I meant you can use that function. 
In your case it would be
```CPP
dtype = internal::GetDefaultDtype(DtypeKind::kInt); 
```

However I'd recommend that you confirm it with the maintainers @niboshi

File: chainerx_cc/chainerx/routines/binary.cc (hunk 1 above)
Comment (kshitij12345) [reply chainerx_cc/chainerx/routines/binary.cc]: Yes looks good to me on a quick glance.

**Comment by sky58** [comment]
Could you review this PR?

File: chainer/_version.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-__version__ = '7.0.0a1'
```
Comment (niboshi) [inline chainer/_version.py]: This change seems unrelated. Please revert it.
(ditto for other similar instances)

Comment (sky58) [reply chainer/_version.py]: I did this.

File: chainer/_version.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
             'cupy-cuda80',
             'cupy',
         ],
-        'specifier': '==7.0.0a1',
+        'specifier': '==7.0.0b1',
```
Comment (niboshi) [inline chainer/_version.py]: Please fix this as well.
It seems you did conflict resolution in a wrong way (there are unrelated commits in the history).
Could you fix that (I guess squash is the easiest way)?

File: chainerx_cc/chainerx/kernels/binary.h [hunk 4]
```diff
@@ -50,4 +50,46 @@ class BitwiseXorASKernel : public Kernel {
     virtual void Call(const Array& x1, Scalar x2, const Array& out) = 0;
 };
 
+class LeftShiftKernel : public Kernel {
```
Comment (niboshi) [inline chainerx_cc/chainerx/kernels/binary.h]: Please name as `LeftShiftAAKernel`.

File: chainerx_cc/chainerx/kernels/binary.h [hunk 5]
```diff
@@ -50,4 +50,46 @@ class BitwiseXorASKernel : public Kernel {
...
+};
+
+class LeftShiftSAKernel : public Kernel {
+public:
+    static const char* name() { return "LeftShiftSA"; }
+
+    virtual void Call(Scalar x1, const Array& x2, const Array& out) = 0;
+};
+
+class RightShiftKernel : public Kernel {
```
Comment (niboshi) [inline chainerx_cc/chainerx/kernels/binary.h]: Ditto. `RightShiftAAKernel`.

File: chainerx_cc/chainerx/dtype.h [hunk 6]
```diff
@@ -207,6 +207,26 @@ auto VisitIntegralDtype(Dtype dtype, F&& f, Args&&... args) {
     }
 }
 
+// Invokes a function by passing PrimitiveType<T> corresponding to given Integral dtype value.
+// See VisitDtype for more detail.
+template <typename F, typename... Args>
+auto VisitShiftDtype(Dtype dtype, F&& f, Args&&... args) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/dtype.h]: It seems we should give better name.

Comment (niboshi) [reply chainerx_cc/chainerx/dtype.h]: Perhaps we should name this as `VisitIntegralDtype` and rename what is currently called `VisitIntegralDtype` to something else.
So, in this PR, could you tentatively write a TODO about that?
Something like "Rename this to VisitIntegralDtype and current VisitIntegralDtype to something else."

File: chainerx/_docs/routines.py [hunk 7]
```diff
@@ -1833,6 +1833,42 @@ def _docs_math():
...
+        chainerx.left_shift,
+        """left_shift(x1, x2)
+Shift the bits of an integer to the left.
+
+Args:
+    x1 (~chainerx.ndarray or scalar): Input array of integers.
+    x2 (~chainerx.ndarray or scalar): Input array of integers.
+
+Returns:
+    :class:`~chainerx.ndarray`: Returned array: :math:`y = x_1 << x_2`
```
Comment (asi1024) [inline chainerx/_docs/routines.py]: In the context of the arithmetic formulae, `x1 << x2` means that `x1` is *much* less than `x2`.

Comment (sky58) [reply chainerx/_docs/routines.py]: So should we change this to `y = x_1 * (2 ^ x_2)` ?

Comment (niboshi) [reply chainerx/_docs/routines.py]: @asi1024 Do you have any alternative? I think no one would be confused.

Comment (asi1024) [reply chainerx/_docs/routines.py]: It seems better to change to `y = x_1 2^{x_2})`, or write in natural language.

Comment (niboshi) [reply chainerx/_docs/routines.py]: I think ` x_1 << x_2` is a lot easier to understand.

Comment (asi1024) [reply chainerx/_docs/routines.py]: Even if so, `x1 << x2` should be written as a code snippet, not as an arithmetic formula.

Comment (niboshi) [reply chainerx/_docs/routines.py]: Hmm, I see. I agree with writing in natural language.

File: chainerx/_docs/routines.py [hunk 8]
```diff
@@ -1833,6 +1833,42 @@ def _docs_math():
...
+        chainerx.left_shift,
+        """left_shift(x1, x2)
+Shift the bits of an integer to the left.
+
+Args:
+    x1 (~chainerx.ndarray or scalar): Input array of integers.
+    x2 (~chainerx.ndarray or scalar): Input array of integers.
+
+Returns:
+    :class:`~chainerx.ndarray`: Return `x1` with bits shifted `x2` times to the left. This is a scalar if both `x1` and `x2` are scalars.
```
Comment (niboshi) [inline chainerx/_docs/routines.py]: >This is a scalar if both `x1` and `x2` are scalars.

I don't think scalar-scalar input is supported.

File: chainerx/_docs/routines.py [hunk 9]
```diff
@@ -1833,6 +1833,42 @@ def _docs_math():
...
+        chainerx.right_shift,
+        """right_shift(x1, x2)
+Shift the bits of an integer to the right.
+
+Args:
+    x1 (~chainerx.ndarray or scalar): Input array of integers.
+    x2 (~chainerx.ndarray or scalar): Input array of integers.
+
+Returns:
+    :class:`~chainerx.ndarray`: Return `x1` with bits shifted `x2` times to the right. This is a scalar if both `x1` and `x2` are scalars.
```
Comment (niboshi) [inline chainerx/_docs/routines.py]: ditto.

**Comment by niboshi** [comment]
Thanks!
Jenkins, test this please

**Comment by niboshi** [comment]
Oh, could you check the error?

**Comment by sky58** [comment]
I fix the error. Could you run Jenkins test again?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Thank you so much!
I'm sorry for taking time.
